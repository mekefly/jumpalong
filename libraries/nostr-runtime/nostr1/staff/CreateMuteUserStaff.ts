import { TYPES } from "../nostr";
import { createStaffFactory, StaffState } from "./Staff1";

export default createStaffFactory()(() => {
  return {
    push(e) {
      const muteListSynchronizer = this.beltline
        .getNostrContainer()
        .get(TYPES.MuteListSynchronizer);

      if (muteListSynchronizer.getMuteList().publicList.has(e.pubkey))
        return StaffState.BREAK;
    },
  };
});
