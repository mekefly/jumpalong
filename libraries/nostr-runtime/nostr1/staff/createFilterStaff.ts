import { Event, Filter } from "nostr-tools";
import { EventBeltline } from "../eventBeltline";
import { EventListFeat } from "./EventStaff/EventListStaff";
import { createStaff, StaffState } from "./Staff1";

export type FilterStaff = {
  push: (event: Event, eventList: Event[]) => StaffState;
};

export function createFilterStaff(filters: Filter[]) {
  return createStaff(
    <FEAT extends EventListFeat>(line: EventBeltline<FEAT>) => {
      line.addPushFilter((line, event) => {
        const eventList = line.feat.getList();
        if (toFilters(event, filters, eventList.length)) {
          return true;
        } else {
          return false;
        }
      });
    }
  );
}

export function toFilters(event: Event, filters: Filter[], length: number) {
  if (filters.length === 0) {
    return true;
  }
  return filters.some((filter) => {
    return toFilter(event, filter, length);
  });
}

export function toFilter(
  event: Event,
  filter: Filter & { search?: string },
  length: number
) {
  const { ids, kinds, authors, since, until, limit, search } = filter;

  if (ids && !ids.includes(event.id as any)) {
    return false;
  } else if (kinds && !kinds.includes(event.kind)) {
    return false;
  } else if (authors && !authors.includes(event.pubkey)) {
    return false;
  } else if (since && !(event.created_at >= since)) {
    return false;
  } else if (until && !(event.created_at <= until)) {
    return false;
  } else if (limit && !(length < limit)) {
    return false;
  } else if (!filterTags(filter, event.tags)) {
    return false;
  } else if (
    search &&
    !JSON.stringify(event).toLowerCase().includes(search.toLowerCase())
  ) {
    return false;
  } else {
    return true;
  }
}

export function filterTags(filter: Filter, tags: string[][]) {
  for (let f in filter) {
    if (f[0] === "#") {
      let tagName = f.slice(1);
      let filterTagValues = filter[`#${tagName}`];
      if (
        filterTagValues &&
        !tags.find(([t, v]) => t === tagName && filterTagValues.includes(v))
      )
        return false;
    }
  }
  return true;
}
