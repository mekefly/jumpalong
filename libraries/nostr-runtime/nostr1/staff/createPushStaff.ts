import { createStaffFactory } from "./Staff1";

export default createStaffFactory()(() => {
  return {
    initialization() {
      this.beltline.feat.pushEvent = function (e, eventList) {
        eventList.push(e);
      };
    },
  };
});
