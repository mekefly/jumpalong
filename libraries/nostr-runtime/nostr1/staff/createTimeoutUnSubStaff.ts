import { createStaffFactory } from "./Staff1";

/**
 * 超时停止订阅
 */
export default createStaffFactory()((timeout: number = 30 * 1000) => {
  return {
    initialization() {
      this.beltline.onAfterReq(({ subId }) => {
        setTimeout(() => {
          this.beltline.closeReqBySubid(subId);
        }, timeout);
      });
    },
  };
});
