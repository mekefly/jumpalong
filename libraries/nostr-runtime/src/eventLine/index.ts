export * from './EventLine'
export * from './types'
