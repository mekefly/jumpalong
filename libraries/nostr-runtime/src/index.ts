export * from './eventLine'
export * from './staff'
export * from './utils'
export * from './eventMap'
export * from './event'
export * from './nostrApi'
export type * from './types'

export * as Synchronizer from './Synchronizer'
