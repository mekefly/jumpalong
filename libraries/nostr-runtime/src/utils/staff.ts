import { EventLineFactory, createStaff } from '..'
export {}


