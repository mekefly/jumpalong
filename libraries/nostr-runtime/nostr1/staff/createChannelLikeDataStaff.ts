import { objectGet } from "@/utils/utils";
import { Event } from "nostr-tools";
import { createStaff, StaffState, type StaffThisType } from ".";
import { EventBeltline } from "r:/Users/meke/Documents/study/jumpalong/src/nostr/eventBeltline";

export const JOIN_CHANNEL_CONTENT = "+";
type ChannelLikeDataStaff = {
  push(this: StaffThisType<{}>, event: Event): StaffState.BREAK;
  feat: {
    channelDataList: Set<ChannelLikeData>;
    getChannelIds(this: { beltline: EventBeltline<{}> }): Set<string>;
    getChannelLikeList(this: {
      beltline: EventBeltline<{}>;
    }): Set<ChannelLikeData>;
  };
};

/**
 * 制作喜欢的频道列表流水线工人
 * @returns
 */
export default function createChannelLikeDataStaff(): ChannelLikeDataStaff {
  return createStaff({
    push(event) {
      if (event.content === JOIN_CHANNEL_CONTENT) {
        objectGet<Set<ChannelLikeData>>(
          this.beltline.feat,
          "channelDataList"
        ).add(createChannelLikeData(event));
      }
      return StaffState.BREAK;
    },
    feat: {
      channelDataList: new Set() as Set<ChannelLikeData>,
      getChannelIds() {
        return new Set(
          Array.from(
            objectGet<Set<ChannelLikeData>>(
              this.beltline.feat,
              "channelDataList"
            ),
            (item) => {
              return item.channelId;
            }
          )
        );
      },
      getChannelLikeList() {
        return objectGet<Set<ChannelLikeData>>(
          this.beltline.feat,
          "channelDataList"
        );
      },
    },
  });
}
function createChannelLikeData(event: Event): ChannelLikeData {
  const data: Partial<ChannelLikeData> = {
    joinId: event.id,
    joinPublisher: event.pubkey,
  };

  for (const tag of event.tags) {
    const t = tag[0];
    if (t === "e") {
      data.channelId = tag[1];
    } else if (t === "p") {
      data.channelCreator = tag[2];
    }
  }

  return data as any;
}

export type ChannelLikeData = {
  joinId: string;
  channelId: string;
  relayUrl?: string;
  channelCreator?: string;
  joinPublisher?: string;
};
