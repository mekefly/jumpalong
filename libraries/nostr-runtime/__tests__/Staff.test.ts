import '@jumpalong/logger'
import { objectFilter } from '@jumpalong/shared'
import { createStaff } from '../src/staff/staff'
import { EventLineFactory } from '../src/eventLine'

describe('staff', () => {
  test('创建Staff应返回一个函数', () => {
    expect(createStaff(line => line.defineEmit<'3'>())).toMatchInlineSnapshot(
      '[Function]'
    )
  })
  test('具有名字', () => {
    expect(
      objectFilter(
        createStaff('id-xx', line => line.defineEmit<'3'>()),
        'id',
        'dependents'
      )
    ).toMatchInlineSnapshot(`
    {
      "dependents": [Function],
      "id": "id-xx",
    }
  `)
  })

  test('具有依赖', () => {
    let xx = createStaff(l => l)
    let yy = createStaff(xx, 'id-xx', line => line.defineEmit<'3'>())
    expect(yy?.dependents?.()).toMatchInlineSnapshot(`
    [
      [Function],
    ]
  `)
    expect(yy.id).toMatchInlineSnapshot(`"id-xx"`)
  })

  test('staff:依赖自动注入', () => {
    //staff1
    let s1 = createStaff(l => {
      return l.assignFeat({ s1: true })
    })

    //staff2
    let s2 = createStaff(s1, l => l.assignFeat({ s2: true }))

    expect(s2).toMatchInlineSnapshot('[Function]')
    let l = new EventLineFactory().add(s2)
    expect(l.out().s1).toMatchInlineSnapshot(`true`)
    expect(l.out().s2).toMatchInlineSnapshot(`true`)
  })
  test('staff:依赖重复注入', () => {
    //staff1
    let n = 0
    let s1 = createStaff(l => {
      //只会执行一次
      expect(n++).toMatchSnapshot()
      return l.assignFeat({ s1: true })
    })

    //staff2
    let s2 = createStaff(s1, s1, l => l.assignFeat({ s2: true }))

    expect(s2).toMatchInlineSnapshot('[Function]')
    let l = new EventLineFactory().add(s2)
    expect(l.out()).toMatchObject({ s1: true, s2: true })
  })

  test('staff:复杂依赖', () => {
    //staff1
    let n = 0
    let s11 = createStaff(function s1(l) {
      return l.assignFeat({ s1: true })
    })

    //staff2
    let s22 = createStaff(s11, function s1(l) {
      return l.assignFeat({ s2: true })
    })
    let s33 = createStaff(s11, function s3(l) {
      return l.assignFeat({ s3: true })
    })
    let s44 = createStaff(s22, function s3(l) {
      return l.assignFeat({ s4: true })
    })
    let s55 = createStaff(s44, function s5(l) {
      return l.assignFeat({ s5: true })
    })
    let s66 = createStaff(s11, s22, s33, s44, s55, function s2(l) {
      return l.assignFeat({ s6: true })
    })

    let l = new EventLineFactory().add(s66)
    expect(l.out()).toMatchObject({
      s1: true,
      s2: true,
      s3: true,
      s4: true,
      s5: true,
      s6: true,
    })
  })

  test('数组依赖注入', () => {
    let s1 = createStaff(m => m.assignFeat({ x1: 1 }))
    let s2 = createStaff([s1], m => m.assignFeat({ x2: 2 }))
    let l = new EventLineFactory().add(s2)
    expect(l.out()).toMatchObject({
      x1: 1,
      x2: 2,
    })
  })

  test('函数依赖注入', () => {
    let s1 = createStaff(m => m.assignFeat({ x1: 1 }))
    let s2 = createStaff(
      () => [s1],
      m => m.assignFeat({ x2: 2 })
    )
    let l = new EventLineFactory().add(s2)
    expect(l.out()).toMatchObject({
      x1: 1,
      x2: 2,
    })
  })
  test('函数依赖注入,具有id', () => {
    let s1 = createStaff(m => m.assignFeat({ x1: 1 }))
    let s2 = createStaff(
      () => [s1],
      'name-test',
      m => m.assignFeat({ x2: 2 })
    )
    expect(s2.id).toMatchInlineSnapshot(`"name-test"`)
    let l = new EventLineFactory().add(s2)
    expect(l.out()).toMatchObject({
      x1: 1,
      x2: 2,
    })
  })
  test('循环依赖', () => {
    let xx = {} as any

    xx.s1 = createStaff(
      () => [xx.s2],
      m => m.assignFeat({ x1: 1 })
    )

    xx.s2 = createStaff(
      () => [xx.s1],
      m => m.assignFeat({ x2: 2 })
    )

    let l = new EventLineFactory().add(xx.s2)

    expect(l.out()).toMatchObject({
      x1: 1,
      x2: 2,
    })
  })
})
