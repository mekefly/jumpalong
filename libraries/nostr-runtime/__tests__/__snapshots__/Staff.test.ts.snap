// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`staff > staff:依赖重复注入 2`] = `0`;
